### ISSUE TITLE
Some questions about training and testing

### ISSUE THREAD

**Issue by YiLiangNie** [Jianlong-Fu/Recurrent-Attention-CNN#7]
when will you open the solver.prototxt file for training and the test when the synset.txt file for cub-200 to test the model?

**Comment by super-wcg** [comment]
@YiLiangNie Hi,did you train the model?

**Comment by chenbinghui1** [comment]
@super-wcg I froze the apn nets, randomly initialized 4 classification layers for 3 scales and a fusion scale, then fine-tune the model based on given one with 1e-4 learning rate, while i only got 81.2% on CUB. If i froze all the apn and conv layers, only trained the classification layers,  i got 83.5% not 85% as in paper. What about you?

**Comment by chenfeima** [comment]
@chenbinghui1 How to do this? I need your help to run the project. 
1. Can this project run in Linux?
2. I have download the dataset "CUB_200_2011". But I don't know what should I do next? 
If you can write a details  doc, Which is much better. My e-mail address is 1691767172@qq.com
I really need you help. Thank you very much!

**Comment by jens25** [comment]
@chenfeima Here is what I did, in order to train the network.
If I get the paper right, the training consists of three steps. I skipped the initialization with the vgg weights and the reinforcement learning part, because reinforcement learning algorithms are not part of caffe. 
So I created three different train_val.prototxt files. 
I used the first one to train the scaling subnets and freezed all other layers.
The second one is used to train the attention proposal networks with a ranking loss to convergence. A sample implementation of this layer can be found here: https://github.com/wanji/caffe-sl/blob/master/src/caffe/layers/pairwise_ranking_loss_layer.cpp
In this second stage all scaling layers are freezed. 
In the final training stage all layers are freezed and the final output layers of the network are trained, which combine the output of the different scalings.
I made a gist, with the training prototxt files belonging to this three training steps. You may use it as a starting point.
https://gist.github.com/jens25/6b0ea1143599fb99bd499a08dd5c072c

**Comment by chenbinghui1** [comment]
@jens25 I have some questions. (1)What's your final results? Does it close to 85%? (2)As shown in your prototxt, the final classifier for each scale is a 100-class classifier? While in cub there are 200 classes.(3)I think directly fine-tune on the given model with freezing all the layers except for the classifier layers, i.e. your stage-3, the result should be close to 85%, while in fact it only has 83%.

**Comment by jens25** [comment]
@chenbinghui1 I don't have any results. I just created the prototxt files, in order to train the network on a custom dataset. I didn't evaluated it on the bird dataset now. Maybe a direct finetuning of the model will give you better results, than this approach.

**Comment by chenfeima** [comment]
@jens25 I download the dataset "CUB_200_2011", but I can not Transform it to lmdb.
Can you give a script to transform?

**Comment by chenfeima** [comment]
@chenbinghui1 I run the test net, also get the result 83%. Do you know why? Have you got the 85%?

**Comment by chenbinghui1** [comment]
@chenfeima If you only test it with the given model, It actually will be 85%. And if you fine-tune it (only fine-tune the classifier layers), you will get 83% and I don't know why.

**Comment by chenfeima** [comment]
@chenbinghui1 I do not fine tune it, But I only get 83%. Whether I downloaded the test net is not the original test net?   Can you give the test net which can get 85%？By the way，I also need the ranking loss and the train net. Thank you very much.
My email address 169767172@qq.com

**Comment by chenfeima** [comment]
@chenbinghui1  Sorry, My email address is 1691767172@qq.com, not 169767172@qq.com

**Comment by chenfeima** [comment]
@jens25 Thank you very much!

**Comment by chenfeima** [comment]
@jens25 I want to know how to Initialize the net ? If only one vgg19, I konw "--weights=caffemodel" But there are 3. I don't konow how to use the same caffemodel to Initialize it?

**Comment by Zyj061** [comment]
@chenfeima Hello, have you solve the initialization problem? I initialized the network by setting weight sharing in the train***.prototxt and saved the model after the network have been initialized before training. Then using this caffe model as my pre-trained model. could anyone tell me whether I'm correct?

**Comment by chenfeima** [comment]
@Zyj061Using the python Interface:1. read caffemodels. 2. copy the params to new caffemodel by layer names.

**Comment by cocowf** [comment]
@chenbinghui1 How can I add attentioncrop layer and rankloss in caffe.proto?I need some help about message parameter compile in caffe.proto

**Comment by lhCheung1991** [comment]
@jens25 
Hi, jens25. I have followed your steps, but it didn't work. The model just never converge.
I skipped the initialization with the VGG weights and used Adam with lr=1e-4. It ran on the CUB_200_2011 for about 10 epochs with loss1, loss2, loss3 floating about 5.

What may cause this situation? Ang suggestion will be appreciated.

**Comment by ouceduxzk** [comment]
@lhCheung1991  are you also trying to reproduce the result? Maybe we can talk together offline

**Comment by lhCheung1991** [comment]
@ouceduxzk 
I am very appreciated for your message. I noticed that you have made some effort on re-implementing this paper. I am looking forward to communicating with you for some details.

**Comment by jackshaw** [comment]
Hi,@chenbinghui1 Could you please tell me how you prepared your test data when testing the pretrained RA_CNN model? I can only get 74% accuracy using the available pretrained model,and I don't know why.

**Comment by ProblemTryer** [comment]
@jackshaw Hi, can you leave a contact to me (maybe qq)?  My email address is honglonghelong@163.com
I have some trouble in start training the model. 
thank you very much!

**Comment by yuqiu1233** [comment]
@chenbinghui1  I'm glad to see you have reimplement this paper in caffe framework.I met some question in the process. I'm very appreciate if you can help me solve the problem.  when i froze the classification layers to train the apn nets, i had met a problem like this:          
        Cannot copy param 0 weights from layer 'get_abc1'; shape mismatch.  Source param shape is 1024 25088 (25690112); target param shape is 1024 100352 (102760448). To learn this layer's parameters from scratch rather than copying from a saved net, rename the layer.

if I change the name of 'get_abc1',the accuracy always 0.98 and the loss always  262.01,the subnet loss like this:Test net output #1: loss1 = 80.42 (* 1 = 80.42 loss),Test net output #2: loss2 = 80.4129 (* 1 = 80.4129 loss), Test net output #3: loss3 = 80.4128 (* 1 = 80.4128 loss).

Do you know what cause this ? and how can i solve the problem?

**Comment by yuqiu1233** [comment]
@lhCheung1991 I met the same problem with you, the loss float arount 5.2. Do you know what cause the problem, and how can solve this problem.  Thank you very much!
